File: Ai_Turtor_FE/src/features/student/chat/understandingCheck.js
const CHECK_HEADING = /^(#{1,6})\s*(?:kiểm tra hiểu|understanding check|câu(?:\s+hỏi)?\s+trắc nghiệm mới(?:\s*\((?:(?:được|đã)\s+)?paraphrase\))?|câu hỏi ôn tập)\s*$/im;
const OPTION_MARK = /(?:^|\s)(?:\(([A-Da-d])\)|([A-Da-d])\.)\s+/g;
const CANONICAL_ANSWER = /(?:^|\s)(?<!chọn\s)(?<!choose\s)(?<!pick\s)(?:đáp án|dap an|(?:the\s+)?(?:correct\s+)?answer)(?:\s+đúng)?\s*(?:là|is|:|：|-)?\s*([A-Da-d])\b[^\n]*/i;
const REVEALED_ANSWER = /(?:^|\s)\s*[-*+]?\s*chọn đáp án đúng\s*[:：-]\s*(?:\n\s*[-*+]?\s*)?([A-Da-d])\s*[.)]?\s*[^\n]*/i;
const ANSWER_PROMPT_MARKER = /(?:^|\s)\s*[-*+]?\s*chọn đáp án đúng\s*[:：-]\s*(?:[-*+]\s*)?/i;
const LEAKED_ANSWER = /nếu bạn chọn(?:\s+đáp án)?\s+([A-Da-d])\b/i;
const CHOOSE_ANSWER = /(?:chọn|choose|pick)\s+(?:đáp án\s+)?([A-Da-d])\b/i;
const LONE_KEY = /(?:^|\n)\s*([A-Da-d])\s*[.)]?\s*$/;
const EXPLAIN_MARKER = /(?:^|\s)(?:giải thích|giai thich|explanation|lý do|ly do)\s*[:：]\s*/ig;
const QUESTION_PREFIX = /^(?:câu hỏi|cau hoi|question)\s*[:：]\s*/im;

function stripDecorations(value) {
  return String(value || '')
    .replace(/^\s*[-*+]\s+/gm, '')
    .replace(/[*_`]+/g, '')
    .replace(/\s+-{2,}\s*$/g, '')
    .replace(/^\s*-{3,}\s*$/gm, '')
    .replace(/\s+/g, ' ')
    .trim();
}

function nextSectionBreak(text) {
  const heading = text.search(/(?:^|\n)#{1,6}\s+\S/);
  return heading >= 0 ? heading : -1;
}

function javaStringHash(value) {
  let hash = 0;
  const text = String(value || '');
  for (let index = 0; index < text.length; index += 1) {
    hash = ((hash * 31) + text.charCodeAt(index)) | 0;
  }
  return hash;
}

function shuffleCorrectAnswerPosition(question, options, correctKey) {
  if (!correctKey || options.length < 2) return { options, correctKey };
  const correctOption = options.find((option) => option.key === correctKey);
  if (!correctOption) return { options, correctKey: '' };

  const seedInput = `${question}\u001f${options
    .map((option) => `${option.key}\u001e${option.text}`)
    .join('\u001f')}`;
  const targetIndex = ((javaStringHash(seedInput) % options.length) + options.length) % options.length;
  const reordered = options.filter((option) => option.key !== correctKey);
  reordered.splice(targetIndex, 0, correctOption);

  let shuffledCorrectKey = '';
  const relabeled = reordered.map((option, index) => {
    const key = String.fromCharCode(65 + index);
    if (option.key === correctKey) shuffledCorrectKey = key;
    return { key, text: option.text };
  });
  return { options: relabeled, correctKey: shuffledCorrectKey };
}

export function normalizeStructuredUnderstandingQuiz(value) {
  if (!value || typeof value !== 'object') return null;
  const question = String(value.question || '').trim();
  const seenKeys = new Set();
  const options = (Array.isArray(value.options) ? value.options : [])
    .map((option) => ({
      key: String(option?.key || '').trim().toUpperCase(),
      text: String(option?.text || '').trim(),
    }))
    .filter((option) => {
      if (!/^[A-D]$/.test(option.key) || !option.text || seenKeys.has(option.key)) return false;
      seenKeys.add(option.key);
      return true;
    });
  if (!question || options.length < 2) return null;

  const requestedKey = String(value.correctKey || '').trim().toUpperCase();
  const correctKey = options.some((option) => option.key === requestedKey) ? requestedKey : '';
  return {
    question,
    options,
    correctKey,
    explanation: String(value.explanation || '').trim(),
  };
}

export function parseUnderstandingQuiz(sectionBody) {
  const raw = String(sectionBody || '')
    .replace(/[*_`]+/g, '')
    .trim();
  if (!raw) return null;

  const canonicalMatch = raw.match(CANONICAL_ANSWER);
  const revealedAnswerMatch = raw.match(REVEALED_ANSWER);
  const leakedMatch = raw.match(LEAKED_ANSWER);
  const chooseMatch = raw.match(CHOOSE_ANSWER);
  const loneMatch = raw.match(LONE_KEY);
  const answerMatch = canonicalMatch || revealedAnswerMatch || leakedMatch || chooseMatch || loneMatch;
  const answerPromptMatch = raw.match(ANSWER_PROMPT_MARKER);
  let answerPromptIndex;
  if (Number.isInteger(answerPromptMatch?.index)) {
    const optionMarkersBeforePrompt = [...raw.slice(0, answerPromptMatch.index).matchAll(OPTION_MARK)];
    if (optionMarkersBeforePrompt.length >= 2) answerPromptIndex = answerPromptMatch.index;
  }
  const explainMarkers = [...raw.matchAll(EXPLAIN_MARKER)];
  const firstExplain = explainMarkers[0];
  const explanationEndCandidates = [
    ...explainMarkers.slice(1).map((item) => item.index),
    canonicalMatch?.index,
    revealedAnswerMatch?.index,
    raw.length,
  ].filter((index) => Number.isInteger(index) && index > (firstExplain?.index ?? -1));
  const explanationEnd = explanationEndCandidates.length
    ? Math.min(...explanationEndCandidates)
    : raw.length;
  const parsedExplanation = firstExplain
    ? raw.slice(firstExplain.index + firstExplain[0].length, explanationEnd)
    : '';
  const metadataIndexes = [
    canonicalMatch?.index,
    revealedAnswerMatch?.index,
    answerPromptIndex,
    ...explainMarkers.map((item) => item.index),
  ].filter(Number.isInteger);
  let working = raw;
  if (metadataIndexes.length) {
    working = working
      .slice(0, Math.min(...metadataIndexes))
      .replace(/(?:^|\n)\s*[-+]\s*$/, '');
  }
  else if (!leakedMatch && !chooseMatch && loneMatch) working = working.replace(loneMatch[0], '\n');
  working = working.replace(QUESTION_PREFIX, '').trim();

  const marks = [];
  OPTION_MARK.lastIndex = 0;
  let match = OPTION_MARK.exec(working);
  while (match) {
    marks.push({
      key: (match[1] || match[2]).toUpperCase(),
      start: match.index,
      textStart: match.index + match[0].length,
    });
    match = OPTION_MARK.exec(working);
  }
  if (marks.length < 2) return null;

  const question = stripDecorations(working.slice(0, marks[0].start))
    .replace(/[?\s]+$/, '');
  if (!question) return null;

  const options = marks.map((item, index) => {
    const end = index + 1 < marks.length ? marks[index + 1].start : working.length;
    return {
      key: item.key,
      text: stripDecorations(working.slice(item.textStart, end)).replace(/[;|]+$/, ''),
    };
  }).filter((item) => item.text);

  if (options.length < 2) return null;

  let correctKey = answerMatch ? answerMatch[1].toUpperCase() : '';
  let explanation = stripDecorations(
    parsedExplanation
      .replace(CANONICAL_ANSWER, ' ')
      .replace(REVEALED_ANSWER, ' ')
      .replace(EXPLAIN_MARKER, ' '),
  );
  const last = options[options.length - 1];
  const leaked = last?.text?.match(
    /^(.{12,}?[.!?])\s+(Nếu bạn chọn(?:\s+đáp án)?\s+[A-D]\b[\s\S]+)$/i,
  );
  if (leaked) {
    last.text = stripDecorations(leaked[1]);
    const leakedKey = leaked[2].match(LEAKED_ANSWER);
    if (!correctKey && leakedKey) correctKey = leakedKey[1].toUpperCase();
    if (!explanation) explanation = stripDecorations(leaked[2]);
  }

  const shuffled = shuffleCorrectAnswerPosition(question, options, correctKey);
  return {
    question: `${question}?`,
    options: shuffled.options,
    correctKey: shuffled.correctKey,
    explanation,
  };
}

export function extractUnderstandingCheck(markdown) {
  const text = String(markdown || '');
  const heading = CHECK_HEADING.exec(text);
  if (!heading) {
    return { before: text, after: '', quiz: null };
  }

  const afterHeading = heading.index + heading[0].length;
  const rest = text.slice(afterHeading).replace(/^\s*\n/, '');
  const breakAt = nextSectionBreak(rest);
  const sectionBody = (breakAt < 0 ? rest : rest.slice(0, breakAt)).trim();
  const quiz = parseUnderstandingQuiz(sectionBody);
  if (!quiz) {
    return { before: text, after: '', quiz: null };
  }

  const after = breakAt < 0 ? '' : rest.slice(breakAt).replace(/^\s*-{3,}\s*/, '').trim();
  return {
    before: text.slice(0, heading.index).trimEnd(),
    after,
    quiz,
  };
}

export function buildUnderstandingCheckPrompt(quiz, selected) {
  const question = String(quiz?.question || '').trim();
  const choice = String(selected?.key || '').trim();
  const choiceText = String(selected?.text || '').trim();
  if (!question || !choice) return '';
  const options = (Array.isArray(quiz.options) ? quiz.options : [])
    .map((item) => `${item.key}. ${item.text}`)
    .join('\n');
  return [
    'Đây là câu kiểm tra hiểu trong bài đang học, không phải chủ đề mới.',
    `Câu hỏi: ${question}`,
    'Lựa chọn:',
    options,
    `Học sinh chọn: ${choice}. ${choiceText}`,
    'Hãy chấm: nói rõ Đúng hay Chưa đúng, nêu đáp án đúng, và giải thích ngắn theo tài liệu. Không mở lộ trình bài mới.',
  ].filter(Boolean).join('\n');
}

export function buildIncorrectAnswerRemediationPrompt(quiz, selected) {
  const question = String(quiz?.question || '').trim();
  const selectedKey = String(selected?.key || '').trim().toUpperCase();
  const selectedText = String(selected?.text || '').trim();
  const correctKey = String(quiz?.correctKey || '').trim().toUpperCase();
  const correctOption = (Array.isArray(quiz?.options) ? quiz.options : [])
    .find((option) => option.key === correctKey);
  if (!question || !selectedKey || !correctKey || selectedKey === correctKey) return '';

  return [
    'Ôn lại sau câu trả lời chưa đúng.',
    `Câu vừa làm: ${question}`,
    `Em đã chọn: ${selectedKey}. ${selectedText}`,
    `Đáp án đúng: ${correctKey}. ${String(correctOption?.text || '').trim()}`,
    quiz?.explanation ? `Lý do trong bài: ${String(quiz.explanation).trim()}` : '',
    'Hãy tự động giảng lại đúng kiến thức này bằng cách dễ hiểu hơn, dùng cả Giải thích khái niệm (Conceptual Explanation) và Minh họa trực quan (Visual Representation).',
    'Sau đó cho em làm lại một câu trắc nghiệm dễ hơn về cùng kiến thức, nhưng phải paraphrase câu hỏi và các lựa chọn; không lặp nguyên văn câu cũ và không chuyển sang chủ đề mới.',
  ].filter(Boolean).join('\n');
}

export function buildMissingAnswerKeyRemediationPrompt(quiz, selected) {
  const question = String(quiz?.question || '').trim();
  const selectedKey = String(selected?.key || '').trim().toUpperCase();
  const selectedText = String(selected?.text || '').trim();
  const options = (Array.isArray(quiz?.options) ? quiz.options : [])
    .map((option) => `${option.key}. ${String(option.text || '').trim()}`)
    .filter((option) => option.length > 3)
    .join('\n');
  if (!question || !selectedKey || !options) return '';

  return [
    'Ôn lại sau câu kiểm tra chưa có đáp án.',
    `Câu vừa làm: ${question}`,
    'Các lựa chọn:',
    options,
    `Em đã chọn: ${selectedKey}. ${selectedText}`,
    'Hãy dựa vào tài liệu môn học để xác định và chấm đáp án em đã chọn, rồi giảng lại kiến thức này bằng Giải thích khái niệm (Conceptual Explanation) và Minh họa trực quan (Visual Representation).',
    'Sau đó cho em làm lại một câu trắc nghiệm dễ hơn về cùng kiến thức nhưng được paraphrase. Câu mới phải có các lựa chọn A, B, C và kết thúc bằng đúng hai dòng: Đáp án: <A hoặc B hoặc C>; Giải thích: <một câu ngắn>.',
  ].join('\n');
}

File: AI_Tutor_MB/ai_tutor/lib/features/ai_tutor/data/understanding_check.dart
class UnderstandingOption {
  const UnderstandingOption({required this.key, required this.text});

  final String key;
  final String text;
}

class UnderstandingQuiz {
  const UnderstandingQuiz({
    required this.question,
    required this.options,
    this.correctKey = '',
    this.explanation = '',
  });

  final String question;
  final List<UnderstandingOption> options;
  final String correctKey;
  final String explanation;

  UnderstandingOption? optionFor(String key) {
    final needle = key.trim().toUpperCase();
    for (final option in options) {
      if (option.key == needle) return option;
    }
    return null;
  }
}

class UnderstandingCheckExtract {
  const UnderstandingCheckExtract({
    required this.before,
    required this.after,
    this.quiz,
  });

  final String before;
  final String after;
  final UnderstandingQuiz? quiz;
}

class UnderstandingCheckAttempt {
  const UnderstandingCheckAttempt({
    required this.quiz,
    required this.selected,
    required this.isCorrect,
  });

  final UnderstandingQuiz quiz;
  final UnderstandingOption selected;
  final bool isCorrect;
}

UnderstandingCheckExtract resolveUnderstandingCheck({
  required String markdown,
  dynamic structured,
}) {
  final extracted = extractUnderstandingCheck(markdown);
  final quiz =
      normalizeStructuredUnderstandingQuiz(structured) ?? extracted.quiz;
  return UnderstandingCheckExtract(
    before: extracted.before,
    after: extracted.after,
    quiz: quiz,
  );
}

final _checkHeading = RegExp(
  r'^(#{1,6})\s*(?:kiểm tra hiểu|understanding check|câu(?:\s+hỏi)?\s+trắc nghiệm mới(?:\s*\((?:(?:được|đã)\s+)?paraphrase\))?|câu hỏi ôn tập)\s*$',
  caseSensitive: false,
  multiLine: true,
);
final _optionMark = RegExp(r'(?:^|\s)(?:\(([A-Da-d])\)|([A-Da-d])\.)\s+');
final _canonicalAnswer = RegExp(
  r'(?:^|\s)(?<!chọn\s)(?<!choose\s)(?<!pick\s)(?:đáp án|dap an|(?:the\s+)?(?:correct\s+)?answer)(?:\s+đúng)?\s*(?:là|is|:|：|-)?\s*([A-Da-d])\b[^\n]*',
  caseSensitive: false,
);
final _revealedAnswer = RegExp(
  r'(?:^|\s)\s*[-*+]?\s*chọn đáp án đúng\s*[:：-]\s*(?:\n\s*[-*+]?\s*)?([A-Da-d])\s*[.)]?\s*[^\n]*',
  caseSensitive: false,
);
final _answerPromptMarker = RegExp(
  r'(?:^|\s)\s*[-*+]?\s*chọn đáp án đúng\s*[:：-]\s*(?:[-*+]\s*)?',
  caseSensitive: false,
);
final _leakedAnswer = RegExp(
  r'nếu bạn chọn(?:\s+đáp án)?\s+([A-Da-d])\b',
  caseSensitive: false,
);
final _chooseAnswer = RegExp(
  r'(?:chọn|choose|pick)\s+(?:đáp án\s+)?([A-Da-d])\b',
  caseSensitive: false,
);
final _loneKey = RegExp(r'(?:^|\n)\s*([A-Da-d])\s*[.)]?\s*$');
final _explainMarker = RegExp(
  r'(?:^|\s)(?:giải thích|giai thich|explanation|lý do|ly do)\s*[:：]\s*',
  caseSensitive: false,
);
final _questionPrefix = RegExp(
  r'^(?:câu hỏi|cau hoi|question)\s*[:：]\s*',
  caseSensitive: false,
  multiLine: true,
);
final _headingBreak = RegExp(r'(?:^|\n)#{1,6}\s+\S');

int javaStringHash(String value) {
  var hash = 0;
  for (final unit in value.codeUnits) {
    hash = (hash * 31 + unit).toSigned(32);
  }
  return hash;
}

({List<UnderstandingOption> options, String correctKey})
shuffleCorrectAnswerPosition(
  String question,
  List<UnderstandingOption> options,
  String correctKey,
) {
  if (correctKey.isEmpty || options.length < 2) {
    return (options: options, correctKey: correctKey);
  }
  final correctOption = options.where((o) => o.key == correctKey).firstOrNull;
  if (correctOption == null) {
    return (options: options, correctKey: '');
  }
  final seedInput =
      '$question\u001f${options.map((o) => '${o.key}\u001e${o.text}').join('\u001f')}';
  final targetIndex =
      ((javaStringHash(seedInput) % options.length) + options.length) %
      options.length;
  final reordered = options.where((o) => o.key != correctKey).toList();
  reordered.insert(targetIndex, correctOption);
  var shuffledCorrectKey = '';
  final relabeled = <UnderstandingOption>[];
  for (var index = 0; index < reordered.length; index++) {
    final key = String.fromCharCode(65 + index);
    if (reordered[index].key == correctKey) shuffledCorrectKey = key;
    relabeled.add(UnderstandingOption(key: key, text: reordered[index].text));
  }
  return (options: relabeled, correctKey: shuffledCorrectKey);
}

String stripDecorations(String value) {
  return value
      .replaceAll(RegExp(r'^\s*[-*+]\s+', multiLine: true), '')
      .replaceAll(RegExp(r'[*_`]+'), '')
      .replaceAll(RegExp(r'\s+-{2,}\s*$'), '')
      .replaceAll(RegExp(r'^\s*-{3,}\s*$', multiLine: true), '')
      .replaceAll(RegExp(r'\s+'), ' ')
      .trim();
}

int nextSectionBreak(String text) {
  final heading = _headingBreak.firstMatch(text);
  return heading?.start ?? -1;
}

UnderstandingQuiz? normalizeStructuredUnderstandingQuiz(dynamic value) {
  if (value is! Map) return null;
  final map = Map<String, dynamic>.from(value);
  final question = map['question']?.toString().trim() ?? '';
  final seenKeys = <String>{};
  final options = <UnderstandingOption>[];
  final rawOptions = map['options'];
  if (rawOptions is List) {
    for (final entry in rawOptions) {
      if (entry is! Map) continue;
      final option = Map<String, dynamic>.from(entry);
      final key = option['key']?.toString().trim().toUpperCase() ?? '';
      final text = option['text']?.toString().trim() ?? '';
      if (!RegExp(r'^[A-D]$').hasMatch(key) ||
          text.isEmpty ||
          seenKeys.contains(key)) {
        continue;
      }
      seenKeys.add(key);
      options.add(UnderstandingOption(key: key, text: text));
    }
  }
  if (question.isEmpty || options.length < 2) return null;
  final requestedKey = map['correctKey']?.toString().trim().toUpperCase() ?? '';
  final correctKey = options.any((o) => o.key == requestedKey)
      ? requestedKey
      : '';
  return UnderstandingQuiz(
    question: question,
    options: options,
    correctKey: correctKey,
    explanation: map['explanation']?.toString().trim() ?? '',
  );
}

UnderstandingQuiz? parseUnderstandingQuiz(String sectionBody) {
  var raw = sectionBody.replaceAll(RegExp(r'[*_`]+'), '').trim();
  if (raw.isEmpty) return null;

  final canonicalMatch = _canonicalAnswer.firstMatch(raw);
  final revealedAnswerMatch = _revealedAnswer.firstMatch(raw);
  final leakedMatch = _leakedAnswer.firstMatch(raw);
  final chooseMatch = _chooseAnswer.firstMatch(raw);
  final loneMatch = _loneKey.firstMatch(raw);
  final answerMatch =
      canonicalMatch ??
      revealedAnswerMatch ??
      leakedMatch ??
      chooseMatch ??
      loneMatch;
  final answerPromptMatch = _answerPromptMarker.firstMatch(raw);
  int? answerPromptIndex;
  if (answerPromptMatch != null) {
    final before = raw.substring(0, answerPromptMatch.start);
    if (_optionMark.allMatches(before).length >= 2) {
      answerPromptIndex = answerPromptMatch.start;
    }
  }
  final explainMarkers = _explainMarker.allMatches(raw).toList();
  final firstExplain = explainMarkers.isEmpty ? null : explainMarkers.first;
  final explanationEndCandidates = <int>[
    ...explainMarkers.skip(1).map((item) => item.start),
    if (canonicalMatch != null) canonicalMatch.start,
    if (revealedAnswerMatch != null) revealedAnswerMatch.start,
    raw.length,
  ].where((index) => firstExplain == null || index > firstExplain.start);
  final explanationEnd = explanationEndCandidates.isEmpty
      ? raw.length
      : explanationEndCandidates.reduce((a, b) => a < b ? a : b);
  final parsedExplanation = firstExplain == null
      ? ''
      : raw.substring(
          firstExplain.start + firstExplain[0]!.length,
          explanationEnd,
        );

  final metadataIndexes = <int>[
    if (canonicalMatch != null) canonicalMatch.start,
    if (revealedAnswerMatch != null) revealedAnswerMatch.start,
    if (answerPromptIndex != null) answerPromptIndex,
    ...explainMarkers.map((item) => item.start),
  ];
  var working = raw;
  if (metadataIndexes.isNotEmpty) {
    final cut = metadataIndexes.reduce((a, b) => a < b ? a : b);
    working = working
        .substring(0, cut)
        .replaceFirst(RegExp(r'(?:^|\n)\s*[-+]\s*$'), '');
  } else if (leakedMatch == null && chooseMatch == null && loneMatch != null) {
    working = working.replaceFirst(loneMatch[0]!, '\n');
  }
  working = working.replaceFirst(_questionPrefix, '').trim();

  final marks = <({String key, int start, int textStart})>[];
  for (final match in _optionMark.allMatches(working)) {
    final key = (match[1] ?? match[2] ?? '').toUpperCase();
    marks.add((
      key: key,
      start: match.start,
      textStart: match.start + match[0]!.length,
    ));
  }
  if (marks.length < 2) return null;

  final question = stripDecorations(
    working.substring(0, marks.first.start),
  ).replaceFirst(RegExp(r'[?\s]+$'), '');
  if (question.isEmpty) return null;

  final options = <UnderstandingOption>[];
  for (var index = 0; index < marks.length; index++) {
    final end = index + 1 < marks.length
        ? marks[index + 1].start
        : working.length;
    final text = stripDecorations(
      working.substring(marks[index].textStart, end),
    ).replaceFirst(RegExp(r'[;|]+$'), '');
    if (text.isNotEmpty) {
      options.add(UnderstandingOption(key: marks[index].key, text: text));
    }
  }
  if (options.length < 2) return null;

  var correctKey = answerMatch == null
      ? ''
      : (answerMatch[1] ?? '').toUpperCase();
  var explanation = stripDecorations(
    parsedExplanation
        .replaceAll(_canonicalAnswer, ' ')
        .replaceAll(_revealedAnswer, ' ')
        .replaceAll(_explainMarker, ' '),
  );
  final last = options.last;
  final leaked = RegExp(
    r'^(.{12,}?[.!?])\s+(Nếu bạn chọn(?:\s+đáp án)?\s+[A-D]\b[\s\S]+)$',
    caseSensitive: false,
  ).firstMatch(last.text);
  if (leaked != null) {
    options[options.length - 1] = UnderstandingOption(
      key: last.key,
      text: stripDecorations(leaked[1] ?? ''),
    );
    final leakedKey = _leakedAnswer.firstMatch(leaked[2] ?? '');
    if (correctKey.isEmpty && leakedKey != null) {
      correctKey = (leakedKey[1] ?? '').toUpperCase();
    }
    if (explanation.isEmpty) explanation = stripDecorations(leaked[2] ?? '');
  }

  final shuffled = shuffleCorrectAnswerPosition(question, options, correctKey);
  return UnderstandingQuiz(
    question: '$question?',
    options: shuffled.options,
    correctKey: shuffled.correctKey,
    explanation: explanation,
  );
}

UnderstandingCheckExtract extractUnderstandingCheck(String markdown) {
  final text = markdown;
  final heading = _checkHeading.firstMatch(text);
  if (heading == null) {
    return UnderstandingCheckExtract(before: text, after: '', quiz: null);
  }
  final afterHeading = heading.end;
  final rest = text.substring(afterHeading).replaceFirst(RegExp(r'^\s*\n'), '');
  final breakAt = nextSectionBreak(rest);
  final sectionBody = (breakAt < 0 ? rest : rest.substring(0, breakAt)).trim();
  final quiz = parseUnderstandingQuiz(sectionBody);
  if (quiz == null) {
    return UnderstandingCheckExtract(before: text, after: '', quiz: null);
  }
  final after = breakAt < 0
      ? ''
      : rest
            .substring(breakAt)
            .replaceFirst(RegExp(r'^\s*-{3,}\s*'), '')
            .trim();
  return UnderstandingCheckExtract(
    before: text.substring(0, heading.start).trimRight(),
    after: after,
    quiz: quiz,
  );
}

String buildUnderstandingCheckPrompt(
  UnderstandingQuiz quiz,
  UnderstandingOption selected,
) {
  final question = quiz.question.trim();
  final choice = selected.key.trim();
  if (question.isEmpty || choice.isEmpty) return '';
  final options = quiz.options
      .map((item) => '${item.key}. ${item.text}')
      .join('\n');
  return [
    'Đây là câu kiểm tra hiểu trong bài đang học, không phải chủ đề mới.',
    'Câu hỏi: $question',
    'Lựa chọn:',
    options,
    'Học sinh chọn: $choice. ${selected.text}',
    'Hãy chấm: nói rõ Đúng hay Chưa đúng, nêu đáp án đúng, và giải thích ngắn theo tài liệu. Không mở lộ trình bài mới.',
  ].join('\n');
}

String buildIncorrectAnswerRemediationPrompt(
  UnderstandingQuiz quiz,
  UnderstandingOption selected,
) {
  final question = quiz.question.trim();
  final selectedKey = selected.key.trim().toUpperCase();
  final correctKey = quiz.correctKey.trim().toUpperCase();
  final correctOption = quiz.optionFor(correctKey);
  if (question.isEmpty ||
      selectedKey.isEmpty ||
      correctKey.isEmpty ||
      selectedKey == correctKey) {
    return '';
  }
  return [
    'Ôn lại sau câu trả lời chưa đúng.',
    'Câu vừa làm: $question',
    'Em đã chọn: $selectedKey. ${selected.text}',
    'Đáp án đúng: $correctKey. ${correctOption?.text.trim() ?? ''}',
    if (quiz.explanation.trim().isNotEmpty)
      'Lý do trong bài: ${quiz.explanation.trim()}',
    'Hãy tự động giảng lại đúng kiến thức này bằng cách dễ hiểu hơn, dùng cả Giải thích khái niệm (Conceptual Explanation) và Minh họa trực quan (Visual Representation).',
    'Sau đó cho em làm lại một câu trắc nghiệm dễ hơn về cùng kiến thức, nhưng phải paraphrase câu hỏi và các lựa chọn; không lặp nguyên văn câu cũ và không chuyển sang chủ đề mới.',
  ].join('\n');
}

String buildMissingAnswerKeyRemediationPrompt(
  UnderstandingQuiz quiz,
  UnderstandingOption selected,
) {
  final question = quiz.question.trim();
  final selectedKey = selected.key.trim().toUpperCase();
  final options = quiz.options
      .map((option) => '${option.key}. ${option.text.trim()}')
      .where((option) => option.length > 3)
      .join('\n');
  if (question.isEmpty || selectedKey.isEmpty || options.isEmpty) return '';
  return [
    'Ôn lại sau câu kiểm tra chưa có đáp án.',
    'Câu vừa làm: $question',
    'Các lựa chọn:',
    options,
    'Em đã chọn: $selectedKey. ${selected.text}',
    'Hãy dựa vào tài liệu môn học để xác định và chấm đáp án em đã chọn, rồi giảng lại kiến thức này bằng Giải thích khái niệm (Conceptual Explanation) và Minh họa trực quan (Visual Representation).',
    'Sau đó cho em làm lại một câu trắc nghiệm dễ hơn về cùng kiến thức nhưng được paraphrase. Câu mới phải có các lựa chọn A, B, C và kết thúc bằng đúng hai dòng: Đáp án: <A hoặc B hoặc C>; Giải thích: <một câu ngắn>.',
  ].join('\n');
}

String understandingCheckStorageKey(String messageId) {
  return 'understanding-check:student:${messageId.trim()}';
}
